core/src/main/scala/chisel3/Data.scala

```diff
@@ -424,7 +432,13 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
     case _: MemTypeBinding[_] => None
   }
 
-  private[chisel3] def topBinding: TopBinding = topBindingOpt.get
+  private[chisel3] def topBinding: TopBinding = {
+    val t = topBindingOpt.get
+    if (_probeInfo.nonEmpty) {
+      require(t.isInstanceOf[PortBinding] || t.isInstanceOf[ProbeBinding], s"Probes must be bound to ports, not ${t}")
+    }
+    t
```

Comment:
This check should be on `topBindingOpt` a few lines above, not `topBinding`. the `Opt` form is the real implementation and this is more of a "getter" if that makes sense.